kaldb/src/main/java/com/slack/kaldb/preprocessor/PreprocessorService.java

```diff
@@ -0,0 +1,285 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The PreprocessorService consumes from multiple upstream topics, applies rate limiting, transforms
+ * the data format, and then writes out the new message to a common downstream topic targeting
+ * specific partitions. The rate limits, and target partitions are read in via the
+ * ServiceMetadataStore, with the upstream topic, downstream topic, and transforms stored in the
+ * service config.
+ */
+public class PreprocessorService extends AbstractIdleService {
+  private static final Logger LOG = LoggerFactory.getLogger(PreprocessorService.class);
+  private static final long MAX_TIME = Long.MAX_VALUE;
+
+  private final ServiceMetadataStore serviceMetadataStore;
+  private final PreprocessorRateLimiter rateLimiter;
+  private final Properties kafkaProperties;
+  private final List<String> upstreamTopics;
+  private final String downstreamTopic;
+  private final String dataTransformer;
+  private final MeterRegistry meterRegistry;
+
+  private KafkaStreams kafkaStreams;
+  private KafkaStreamsMetrics kafkaStreamsMetrics;
+
+  private final Counter configReloadCounter;
+  public static final String CONFIG_RELOAD_COUNTER = "preprocessor_config_reload_counter";
+
+  public PreprocessorService(
+      ServiceMetadataStore serviceMetadataStore,
+      KaldbConfigs.PreprocessorConfig preprocessorConfig,
+      MeterRegistry meterRegistry) {
+    this.serviceMetadataStore = serviceMetadataStore;
+    this.meterRegistry = meterRegistry;
+    this.configReloadCounter = meterRegistry.counter(CONFIG_RELOAD_COUNTER);
+
+    this.kafkaProperties = makeKafkaStreamsProps(preprocessorConfig.getKafkaStreamConfig());
+    this.downstreamTopic = preprocessorConfig.getDownstreamTopic();
+    this.upstreamTopics = Collections.unmodifiableList(preprocessorConfig.getUpstreamTopicsList());
+    this.dataTransformer = preprocessorConfig.getDataTransformer();
+    this.rateLimiter =
+        new PreprocessorRateLimiter(
+            meterRegistry, preprocessorConfig.getPreprocessorInstanceCount());
+  }
+
+  @Override
+  protected void startUp() throws Exception {
+    LOG.info("Starting preprocessor service");
+    load();
+    serviceMetadataStore.addListener(this::load);
+    LOG.info("Preprocessor service started");
+  }
+
+  @Override
+  protected void shutDown() throws Exception {
+    LOG.info("Stopping preprocessor service");
+    if (kafkaStreams != null) {
+      kafkaStreams.close(DEFAULT_START_STOP_DURATION);
+    }
+    if (kafkaStreamsMetrics != null) {
+      kafkaStreamsMetrics.close();
+    }
+    serviceMetadataStore.removeListener(this::load);
+    LOG.info("Preprocessor service closed");
+  }
+
+  /**
+   * Configures and starts a KafkaStream processor, based off of the cached ServiceMetadataStore.
+   * This method is reentrant, and will restart any existing KafkaStream processors. Access to this
+   * must be synchronized if using this method as part of a listener.
+   */
```

Comment:
What happens to the pre-processor in case of pipeline errors? Would the process die? It would be valuable to document those cases here or in the classdoc.

Reply:
Documentation added describing the topology rebuild. Also slightly tweaked the load method to better catch exceptions in followup loads, and documented what to expect in those situations.